lib/utils/fs.go

```diff
@@ -106,6 +107,9 @@ func OpenFile(path string) (*os.File, error) {
 	if err != nil {
 		return nil, trace.ConvertSystemError(err)
 	}
+	if !allowSymlink && fi.Mode().Type()&os.ModeSymlink != 0 {
+		return nil, trace.BadParameter("symlink not allowed")
+	}
```

Comment:
A couple notes from things I learned trying to do this for `tbot`:
1. This check needs to use `os.Lstat()` to ever return `true`, since `os.Stat()` follows symlinks. We're just reading the destination file info here, which can never be a symlink, even if broken.
2. This one depends on our threat model, but even `os.Lstat()` follows symlinks, if they occur in any preceding path component. It's also a bit hairy to recursively check each path component since there could be cycles and other shenanigans.
   * This caveat applies to any solutions claiming to prevent symlink attacks using some variation of `O_NOFOLLOW`, since that flag only ever applies to the final pathname component.

We use [`openat2()` with `RESOLVE_NO_SYMLINKS`](https://man7.org/linux/man-pages/man2/openat2.2.html) in tbot to prevent this sort of file redirection attack: https://github.com/gravitational/teleport/blob/master/lib/tbot/botfs/fs_linux.go#L62

Unfortunately this comes with some pretty big caveats: `openat2()` isn't supported on older kernels, and is very frequently disabled on container engines and various enterprise Linuxes. And it's Linux-only :/. It's okay as an incremental security improvement where available, though - tbot checks for `openat2()` availability and falls back to insecure writes.

Reply:
It also looks like we evaluate symlinks ahead of time in `NormalizePath()`, so we probably need to be careful with its symlink-following behavior too :slightly_frowning_face:

Reply:
Thank you @timothyb89, this is very helpful.  Let me look at this closer

Reply:
That said, I am surprised to see an OS compatibility issue with `Lstat`: https://github.com/gravitational/teleport/actions/runs/6150975451

Should this be os conditional?

Reply:
I see the issue, working on a fix now

Reply:
I'm pretty sure that covers it. I've been feeding it inputs and haven't been able to break it, so that's something :slightly_smiling_face:

Reply:
Same!  I have been mentally weighing if I should build unit tests to cover these cases.  The downside is building the directory structure for testing is a fair bit of complexity.  Let me know your thoughts